common.sc

```diff
@@ -0,0 +1,126 @@
+  val version = sys.env.getOrElse(dep + "Version", defaultVersions(dep))
+  if (cross)
+    ivy"$org:::$dep:$version"
+  else
+    ivy"$org::$dep:$version"
+}
+
+/** The reason to split build.sc to two file is
+  * [[CommonRocketChip]] doesn't need to import `$file.chisel3` and `$file.firrtl`.
+  *
+  * You can extends from [[CommonRocketChip]] to use rocket-chip as build-from-source dependency.
+  * When doing this, you may like to override `chisel3Module`, `hardfloatModule`, `configModule`,
+  * setting them to your favorite commit of those packages.
+  *
+  * If you don't override `chisel3Module`, which will default to be `None`,
+  * and mill will automatically use chisel3 from ivy.
+  */
+trait CommonRocketChip extends SbtModule with PublishModule {
+  m =>
+
+  object macros extends SbtModule with PublishModule {
+    override def scalaVersion = T {
+      m.scalaVersion()
+    }
+
+    override def ivyDeps = T {
+      m.ivyDeps()
+    }
+
+    override def pomSettings = T {
+      m.pomSettings()
+    }
+
+    override def publishVersion = T {
+      m.publishVersion()
+    }
+  }
+
+  object test extends Tests {
+    override def scalacPluginClasspath = m.scalacPluginClasspath
+
+    override def ivyDeps = m.ivyDeps() ++ Agg(
+      ivy"org.scalatest::scalatest:3.2.0"
+    )
+
+    def testFrameworks = T {
+      Seq("org.scalatest.tools.Framework")
+    }
+
+    // a sbt-like testOnly command.
+    // for example, mill -i "chisel3[2.12.12].test.testOnly" "chiselTests.BitwiseOpsSpec"
+    def testOnly(args: String*) = T.command {
+      super.runMain("org.scalatest.run", args: _*)
+    }
+  }
+
+  override def millSourcePath = super.millSourcePath / os.up
+
+  def chisel3Module: Option[PublishModule] = None
+
+  def hardfloatModule: PublishModule
+
+  def configModule: PublishModule
+
+  def chisel3IvyDeps = if (chisel3Module.isEmpty) Agg(
+    getVersion("chisel3")
+  ) else Agg.empty[Dep]
+
+  override def mainClass = T {
+    Some("freechips.rocketchip.system.Generator")
+  }
+
+  override def moduleDeps = Seq(macros) ++ chisel3Module :+ hardfloatModule :+ configModule
+
+  override def scalacOptions = T {
+    Seq("-deprecation", "-unchecked", "-Xsource:2.11")
+  }
+
+  override def ivyDeps = T {
+    Agg(
+      ivy"${scalaOrganization()}:scala-reflect:${scalaVersion()}",
+      ivy"org.json4s::json4s-jackson:3.6.1",
+      ivy"org.scalatest::scalatest:3.2.0"
+    ) ++ chisel3IvyDeps
+  }
+
+  private val macroParadise = ivy"org.scalamacros:::paradise:2.1.1"
+
+  private val chisel3Plugin = getVersion("chisel3-plugin")
+
+  override def repositories = super.repositories ++ Seq(
+    MavenRepository("https://oss.sonatype.org/content/repositories/snapshots"),
+    MavenRepository("https://oss.sonatype.org/content/repositories/releases")
+  )
+
+  override def compileIvyDeps = Agg(macroParadise) ++ (if(chisel3Module.isDefined) Agg() else Agg(chisel3Plugin))
```

Comment:
Is this needed since the plugin is declared in the `scalacPluginIvyDeps`?

Reply:
I just checked in:
https://github.com/lihaoyi/mill/blob/cc0d0fc2de175d295e7905f11d709ca69536e616/integration/test/resources/play-json/build.sc#L70-L75, so I think `macroParadise` should exist in both cases.
But I think `chisel3Plugin` only needed in `scalacPluginIvyDeps`.